polar-core/src/data_filtering.rs

```diff
@@ -0,0 +1,799 @@
+                request.constraints.push(Constraint {
+                    kind: ConstraintKind::Eq,
+                    field: field.clone(),
+                    value: ConstraintValue::Field(eqf.1.clone()),
+                });
+                contributed_constraints = true;
+            }
+            assert!(contributed_constraints);
+        }
+    }
+
+    // Constrain any vars that are `in` this var.
+    // Add their constraints to this one.
+    // @NOTE(steve): I think this is right, but I'm not totally sure.
+    // This might assume that the current var is a relationship of kind "children".
+    for (lhs, rhs) in &vars.in_relationships {
+        if rhs == var_id {
+            constrain_var(result_set, types, vars, lhs, var_type);
+            let in_result_set = result_set.requests.remove(lhs).unwrap();
+            assert_eq!(result_set.resolve_order.pop(), Some(lhs.to_string()));
+            request.constraints.extend(in_result_set.constraints);
+        }
+    }
+
+    result_set.requests.insert(var_id.to_string(), request);
+    result_set.resolve_order.push(var_id.to_string());
+}
+
+pub fn opt_pass(filter_plan: &mut FilterPlan, explain: bool) -> bool {
+    let mut optimized = false;
+
+    // Remove duplicate result set in a union.
+    let mut drop_plan = None;
+    'plans: for (i, result_set_a) in filter_plan.result_sets.iter().enumerate() {
+        for (j, result_set_b) in filter_plan.result_sets.iter().enumerate() {
+            if i != j && result_set_a == result_set_b {
+                drop_plan = Some(j);
+                break 'plans;
+            }
+        }
+    }
+    if let Some(plan_id) = drop_plan {
+        if explain {
+            eprintln!("* Removed duplicate result set.")
+        }
+        filter_plan.result_sets.remove(plan_id);
+        optimized = true;
+    }
+
+    // Possible future optimization ideas.
+    // * If two result sets are almost the same except for a single fetch
+    //   that only has a single field check and the field is different, we
+    //   can merge the two result sets and turn the field check into an `in`.
+    //   This is basically "un-expanding" either an `in` or and `or` from the policy.
+    //   This could be hard to find.
+    optimized
+}
+
+pub fn optimize(mut filter_plan: FilterPlan, explain: bool) -> FilterPlan {
+    if explain {
+        eprintln!("\nOptimizing...")
+    }
+    while opt_pass(&mut filter_plan, explain) {}
+    if explain {
+        eprintln!("Done\n")
+    }
+    filter_plan
+}
+
+pub fn build_filter_plan(
+    types: Types,
+    partial_results: PartialResults,
+    variable: &str,
+    class_tag: &str,
+) -> PolarResult<FilterPlan> {
+    let mut filter_plan = FilterPlan {
+        result_sets: vec![],
+    };
+    // @NOTE(steve): Just reading an env var here sucks (see all the stuff we had to do
+    // to get POLAR_LOG to work in all libs, wasm etc...) but that's what I'm doing today.
+    // At some point surface this info better.
```

Comment:
+1. It's probably best for us to pass in all these options & read the env vars from the host libraries. If you want you could open a task for this & mark as good for onboarding & cleanup.

Reply:
Other aspect of this is having various logging we can turn on and off. See the DEBUG constant at the top of simplifier for another crappy example of this that I added.

Reply:
yeah, that is a good onboarding task